plugin/restore/mysql/mysql.go

```diff
@@ -680,6 +680,189 @@ func (r *Restore) GetSortedBinlogFilesMetaOnServer(ctx context.Context) ([]Binlo
+	fields := strings.Fields(line)
+	if len(fields) != 3 {
+		return 0, false, fmt.Errorf("unexpected mysqlbinlog output line %q when parsing binlog event start position", line)
+	}
+	pos, err = strconv.ParseInt(fields[2], 10, 0)
+	if err != nil {
+		return 0, false, err
+	}
+	return pos, true, nil
+}
+
+// Parse the first binlog eventTs of a local binlog file.
+func (r *Restore) parseLocalBinlogFirstEventTs(ctx context.Context, fileName string) (eventTs int64, err error) {
+	args := []string{
+		// Local binlog file path.
+		path.Join(r.binlogDir, fileName),
+		// Tell mysqlbinlog to suppress the BINLOG statements for row events, which reduces the unneeded output.
+		"--base64-output=DECODE-ROWS",
+	}
+	cmd := exec.CommandContext(ctx, r.mysqlutil.GetPath(mysqlutil.MySQLBinlog), args...)
+	cmd.Stderr = os.Stderr
+	pr, err := cmd.StdoutPipe()
+	if err != nil {
+		return 0, err
+	}
+	s := bufio.NewScanner(pr)
+	if err := cmd.Start(); err != nil {
+		return 0, err
+	}
+	defer func() {
+		pr.Close()
+		err = cmd.Process.Kill()
+	}()
+
+	for s.Scan() {
+		line := s.Text()
+		eventTsParsed, found, err := parseBinlogEventTsInLine(line)
+		if err != nil {
+			return 0, fmt.Errorf("failed to parse binlog eventTs from mysqlbinlog output, error: %w", err)
+		}
+		if !found {
+			continue
+		}
+		eventTs = eventTsParsed
+		break
+	}
+
+	return eventTs, nil
+}
+
+// Use command like mysqlbinlog --start-datetime=targetTs binlog.000001 to parse the first binlog event position with timestamp equal or after targetTs.
+// TODO(dragonly): Add integration test.
+func (r *Restore) getBinlogEventPositionAtOrAfterTs(ctx context.Context, binlogFile BinlogFile, targetTs int64, isLastBinlogFile bool) (pos int64, err error) {
+	args := []string{
+		// Local binlog file path.
+		path.Join(r.binlogDir, binlogFile.Name),
+		// Tell mysqlbinlog to suppress the BINLOG statements for row events, which reduces the unneeded output.
+		"--base64-output=DECODE-ROWS",
+		// Instruct mysqlbinlog to start output only after encountering the first binlog event with timestamp equal or after targetTs.
+		"--start-datetime", formatDateTime(targetTs),
+	}
+	cmd := exec.CommandContext(ctx, r.mysqlutil.GetPath(mysqlutil.MySQLBinlog), args...)
+	cmd.Stderr = os.Stderr
+	pr, err := cmd.StdoutPipe()
+	if err != nil {
+		return 0, err
+	}
+	s := bufio.NewScanner(pr)
+	if err := cmd.Start(); err != nil {
+		return 0, err
+	}
+	defer func() {
+		pr.Close()
+		err = cmd.Process.Kill()
+	}()
+
+	var foundPos bool
+	for s.Scan() {
+		line := s.Text()
+		posParsed, found, err := parseBinlogEventPosInLine(line)
+		if err != nil {
+			return 0, fmt.Errorf("failed to parse binlog event start position from mysqlbinlog output, error: %w", err)
+		}
+		if !found {
+			continue
+		}
+		if posParsed == 4 {
+			// When invoking mysqlbinlog with --start-datetime, the first valid event will always be FORMAT_DESCRIPTION_EVENT which should be skipped.
+			continue
+		}
+		pos = posParsed
+		foundPos = true
```

Comment:
foundPos is not needed if we look at pos != 0.

Reply:
SG.